### PR TITLE
add auto-tuning for max_connections

### PR DESCRIPTION
**Pull request by delgod**
## Issue

MySQL `max_connections` is neither being tunned nor being exposed to configuration, staying at the default 151 connections.

## Solution

Add sane auto-tune strategy while proper configuration management is not implemented.

### CODE REVIEW

File: lib/charms/mysql/v0/mysql.py [hunk 1]
```diff
@@ -1513,9 +1514,29 @@ def get_innodb_buffer_pool_parameters(self) -> Tuple[int, Optional[int]]:
...
+        except Exception:
+            logger.exception("Failed to compute innodb buffer pool parameters")
+            raise MySQLGetAutoTunningParametersError("Error computing buffer pool parameters")
+
+    def get_max_connections(self) -> int:
+        """Calculate max_connections parameter for the instance."""
+        # Reference: based off xtradb-cluster-operator
+        # https://github.com/percona/percona-xtradb-cluster-operator/blob/main/pkg/pxc/app/config/autotune.go#L61-L70
+
+        bytes_per_connection = 12582912  # 12 Megabytes
```
Comment (paulomach) [inline lib/charms/mysql/v0/mysql.py:1526]: I wonder if this is way too conservative and can still not be enough.

**Review by taurus-forever — APPROVED** [review]
Just for the history: this will decrease the current max_connections in CI tests.

We allowed 111 connections in the past, but currently we constraint RAM to 1GB, which means ~85 max connections allowed (42? connections for RAM=500Mb installations).

**Review by taurus-forever — APPROVED** [review]
Just for the history: this will decrease the current max_connections in CI tests.

We allowed 111 connections in the past, but currently we constraint RAM to 1GB, which means ~85 max connections allowed (42? connections for RAM=500Mb installations).

File: lib/charms/mysql/v0/mysql.py (hunk 1 above)
Comment (delgod) [reply lib/charms/mysql/v0/mysql.py:1526]: Absolutely conservative, it is sort of "safe defaults".
The idea is to provide safe default, and if the user knows the biggest value that does not produce OOM for his particular workload, then he will customize the value via "juju config" (when we have config management).

**Pull request merged**